pathd/path_pcep_controller.c

```diff
@@ -650,18 +727,62 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
 	ctrl_state->pcc_opts = opts;
 	return 0;
 }
-
+int pcep_thread_event_update_best_pce(struct ctrl_state *ctrl_state, int best)
+{
+	PCEP_DEBUG(" recalculating pce precedence ");
+	if (best) {
+		struct pcc_state *best_pcc_state =
+			get_pcc_state(ctrl_state, best);
+		if (best_pcc_state->pce_opts->previous_best
+		    != best_pcc_state->pce_opts->is_best) {
+			PCEP_DEBUG(" %s Resynchro best (%i) previous best (%i)",
+				   best_pcc_state->tag, best_pcc_state->id,
+				   best_pcc_state->pce_opts->previous_best);
+			best_pcc_state->status = PCEP_PCC_SYNCHRONIZING;
+			best_pcc_state->retry_count = 0;
+			best_pcc_state->synchronized = false;
+			pcep_thread_start_sync(ctrl_state, best_pcc_state->id);
+		} else {
+			PCEP_DEBUG(
+				" %s No Resynchro best (%i) previous best (%i)",
+				best_pcc_state->tag, best_pcc_state->id,
+				best_pcc_state->pce_opts->previous_best);
+		}
+	} else {
+		PCEP_DEBUG(" None best pce , all pce seem disconnected");
+	}
+	return 0;
+}
 int pcep_thread_event_update_pce_options(struct ctrl_state *ctrl_state,
 					 int pcc_id, struct pce_opts *pce_opts)
 {
+	if (!pce_opts || !ctrl_state) {
+		return 0;
+	}
 	struct pcc_state *pcc_state;
 	struct pcc_opts *pcc_opts;
 
+	struct pce_opts *pce_opts_copy;
+	int current_pcc_id = get_pcc_id_by_ip(pcep_g->fpt, &pce_opts->addr);
+	if (!current_pcc_id) {
+		current_pcc_id = pcep_ctrl_get_free_pcc_id(pcep_g->fpt);
+	}
+	if (pcep_g->pce_opts[current_pcc_id - 1] != NULL)
+		XFREE(MTYPE_PCEP, pcep_g->pce_opts[0]);
+	pce_opts_copy = XCALLOC(MTYPE_PCEP, sizeof(*pce_opts));
+	pce_opts_copy = memcpy(pce_opts_copy, pce_opts, sizeof(*pce_opts));
+	pcep_g->pce_opts[current_pcc_id - 1] = pce_opts_copy;
```

Comment:
You should not do that from the controller thread and read it from the main thread without a mutex. This is why this was done in pathd_pcep.c.
In general, pcep_g should only used from the main thread.

Reply:
Yes, back to pathd_pcep.c as originally was